src/librustc_typeck/check/intrinsic.rs

```diff
@@ -382,6 +382,12 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem) {
                 (1, vec![ tcx.mk_mut_ptr(param(0)), param(0) ], tcx.mk_unit())
             }
 
+            "miri_start_panic" => {
+                // FIXME - the relevant types aren't lang items,
+                // so it's not trivial to check this
+                return;
+            }
```

Comment:
Hm, this is suboptimal. I'd like someone from the compiler team to sign this off... @oli-obk ?

Reply:
Yeah, I agree that this is not great. I didn't want to make `Any` and `Send` lang items just to support type-checking this.

I didn't consider it too big of a deal, because:

1) Any user-supplied definition of this intrinsic is completely unsupported, since this is an implementation detail.
2) If we somehow were to change the definition of this intrinsic to something else, it would get caught when we ran the Miri testsuite.

Of course, ideally we would type-check this intrinsic definition like any other.

Reply:
Since this is just for diagnostics, you can use diagnostic items instead of lang items. We'll need to make these diagnostic items at some point anyway for clippy.

Reply:
This is an intrinsic, not a lang item... and it's not just for diagnostic, it is a core part of our proposed panicking strategy for Miri.

Reply:
Why not rely on the way `libstd` communicates with the unwinding runtime instead? I think it's through a hardcoded symbol name, although it could be a weak lang item instead.

Reply:
@eddyb: @Aaron1011 tried that first, it's a mess. The problem is that *after* that hardocded symbol name, the panic run-time does some work, and doing that work "by hand" in Miri is really awful. Basically, the hardcoded symbol name is too early, not low enough of an abstraction.

Here's that common code:
https://github.com/rust-lang/rust/blob/2e4da3caadc61fab2cfcffebcbfdd72fbcee62b7/src/libpanic_unwind/lib.rs#L92-L95

What we want is intercept the call to that `panic::imp`, *after* that `Box::from_raw((*payload).box_me_up())` stuff.

Reply:
Do you want this unwinding behavior in miri to depend on whether `-C panic=abort` is used? Because if you require `panic_unwind` to be in use for miri unwinding, you can probably make that `imp::panic` a lang item or something to intercept it directly.

Or maybe you can just add a lang item for helping miri preprocess the payload?

Reply:
> Because if you require panic_unwind to be in use for miri unwinding, you can probably make that imp::panic a lang item or something to intercept it directly.

This is essentially what we're doing - except the `cfg(miri)` implementation of `imp::panic` just calls the `miri_start_panic`. This is the only case that needs compiler support - all of the other unwinding backends can just be regular functions.

Reply:
I would prefer if `cfg(miri)` and an intrinsic weren't needed here, is my point.

Reply:
I think it's clearer to have `libpanic_unwind` explicitly indicate that differnet logic is being run for Miri. With your approach, we would appear to invoke `imp::panic` for the host platform (e.g. from `src/libpanic_unwind/gcc.rs`), but the call would actually get intercepted by Miri.

Reply:
@eddyb I don't really understand your proposal. We use an intrinsic instead of a lang item, but other than that I think we are doing what you suggest... except you say you wouldn't need `cfg(miri)`, so now I just don't know what it is you mean. Could you be a bit more concrete?

One thing I could imagine is shifting the abstraction. Right now, libcore calls `__rust_start_panic`, and then `__rust_start_panic` does a bit of work before dispatching to something platform-specific. We could move that work to the libcore side, changing the type of `__rust_start_panic` to take `Box<dyn Any + Send>` (return type stays the same).
Miri has to intercept and forward calls to `__rust_start_panic` anyway; with our current proposal it forwards them to the real `panic_unwind` only to very immediately intercept again. So with this alternative proposal, the interception of `__rust_start_panic` is already at the right place. We wouldn't even need `panic_unwind` at all in Miri then.

Reply:
I opened an issue for further  discussion: https://github.com/rust-lang/rust/issues/66247

Thus, marking as resolved here.